src/sql/expr.rs

```diff
@@ -1344,13 +1344,38 @@ impl ScalarExpr {
         }
     }
 
-    pub fn call_unary(self, func: UnaryFunc) -> Self {
+    pub fn call_unary(mut self, func: UnaryFunc) -> Self {
+        #[allow(clippy::single_match)]
+        match func {
+            UnaryFunc::Not => match &mut self {
+                ScalarExpr::CallBinary { expr1, expr2, func } => {
+                    if let Some(negated_func) = negate(&func) {
+                        return ScalarExpr::CallBinary {
+                            expr1: Box::new(expr1.take()),
+                            expr2: Box::new(expr2.take()),
+                            func: negated_func,
+                        };
+                    }
+                }
+                ScalarExpr::CallUnary {
+                    expr,
+                    func: UnaryFunc::Not,
+                } => return expr.take(),
+                _ => {}
+            },
+            _ => {}
+        }
+
         ScalarExpr::CallUnary {
             func,
             expr: Box::new(self),
         }
     }
 
+    pub fn take(&mut self) -> Self {
+        mem::replace(self, ScalarExpr::literal_null(ScalarType::Unknown))
+    }
+
```

Comment:
For `RelationExpr` there are two `take` variants, one of which leaves properly typed data behind and one of which does not. It's probably worth either doing the same here or documenting this one as leaving `self` in a broken state.

Reply:
Can do! But to clarify, it's not my intention to merge this PR—I think there's probably some more broad restructuring that would make using this pattern more appropriate (as it is now, inlining these rules directly in the constructors is a little ugly, and not the right place for that logic to live), I just wanted to see if people had thoughts on the general strategy of intercepting construction like this.

Reply:
Yes, I think a few of us (ping @jamii) prefer strongly normalized representations. Ideally we delete all of the redundant functions and just go down to some sane minimal core. This would make things like CSE more effective, I would imagine. But fair enough about just workshopping it!